### ISSUE TITLE
Support for assertions in regex

### ISSUE THREAD

**Issue by leotrs** [tree-sitter/tree-sitter#1970]
All I could find about this was this issue in an archived repo.

Is writing an external scanner still the suggested workaround to this? If so, I will note that writing an external scanner is a much higher barrier to using tree-sitter as it requires (at least some) C/C++ knowledge, as opposed to simply using a `^`, or `$` in a regex which is standard.

Lookarounds would also be nice to have, but lacking `^` and `$` support is a bit of a headache tbh.

Thanks for all the hard work!

**Comment by ahelwer** [comment]
Could you put `\r?\n` at the beginning of the regex? Assuming it wasn't consumed by the prior line.

**Comment by leotrs** [comment]
Yes I have been able to find simple workarounds for `^` and `$` but not for lookarounds.

The problem is precisely what you mention -- this will consume some whitespace so all other regexes in the grammar will have to be aware of such a fact. So this workaround works fine for a single (or a few) regexes, but when many regexes all have to be aware of each other this starts to become a bit of a nightmare :sweat_smile:

**Comment by ahelwer** [comment]
Yeah. Having non-consuming match groups would be nice. Lookbehind seems difficult to implement but lookahead seems simple enough using the scanner's mark_end functionality. Maybe simple enough for me to implement in a PR, now that I think of it...

**Comment by leotrs** [comment]
Update: I tried implementing a sort of lookaround-only regex by using an external scanner. Essentially, a rule of the form `ends_with_a_period: $ => seq(/[^.]+/, $.period)`, with an external scanner that performs the check `lexer->lookahead == "."`.

Note the entire point of lookaround-like behavior is that it should not consume any characters.

However, this is unfortunately impossible at this time (or at least it's not this simple...) because of [this check](https://github.com/tree-sitter/tree-sitter/blob/36b5b6c89e55ad1a502f8b3234bb3e12ec83a5da/lib/src/parser.c#L456-L459): if the external scanner returns a token with length zero, it is ignored.

Since an external scanner must always consume characters, this kind of simple workaround for lookaround behavior is not possible.

I think there is the possibility of having a scanner with state, and intentionally changing the state of the scanner when the lookaround is successful, if only to bypass the mentioned check. But this honestly sounds a *tad* too complicated simply for checking if the point is behind a period character without consuming it!

**Comment by ahelwer** [comment]
Move the preceding regex into the external scanner. Consume it then use mark_end to look ahead.

**Comment by leotrs** [comment]
Ah yes! I am actually using that workaround for some simple cases.

However, when the middle rule is more complicated i.e. `ends_with_a_period: $ => seq($.some_other_rule, $.period)` this is no longer viable unless I start moving a considerable amount of the language specification into the external scanner.

**Comment by ahelwer** [comment]
Ref https://github.com/tree-sitter/tree-sitter/discussions/1252 for regex lookahead; I'm idly scoping out how difficult this would be to implement today.

**Comment by ObserverOfTime** [comment]
Supporting anchors might be possible, but look-arounds would likely be too complicated.
